test/jasmine/tests/transform_aggregate_test.js

```diff
@@ -0,0 +1,190 @@
+        Plotly.newPlot(gd, [{
+            x: [1, 2, 3, 4, 'fail'],
+            y: [1.1, 2.2, 3.3, 'nope', 5.5],
+            marker: {
+                size: ['2001-01-01', 0.2, 0.1, 0.4, 0.5],
+                color: [2, 4, '', 10, 8],
+                opacity: [0.6, 'boo', 0.2, 0.8, 1.0],
+                line: {
+                    color: [2.2, 3.3, 4.4, 5.5, 'the end']
+                }
+            },
+            transforms: [{
+                type: 'aggregate',
+                groups: ['a', 'b', 'a', 'a', 'a'],
+                aggregations: [
+                    // missing array - the entry is ignored
+                    {array: '', func: 'avg'},
+                    {array: 'x', func: 'sum'},
+                    // non-numerics will not count toward numerator or denominator for avg
+                    {array: 'y', func: 'avg'},
+                    {array: 'marker.size', func: 'min'},
+                    {array: 'marker.color', func: 'max'},
+                    // marker.opacity doesn't have an entry, but it will default to first
+                    // as if it were {array: 'marker.opacity', func: 'first'},
+                    {array: 'marker.line.color', func: 'last'},
+                    // not present in data, but that's OK for count
+                    {array: 'marker.line.width', func: 'count'},
+                    // duplicate entry - discarded
+                    {array: 'x', func: 'min'}
+                ]
+            }]
+        }], {
+            // log axis doesn't change how sum (or avg but not tested) works
+            xaxis: {type: 'log'}
+        });
+
+        var traceOut = gd._fullData[0];
+
+        expect(traceOut.x).toEqual([8, 2]);
+        expect(traceOut.y).toBeCloseToArray([3.3, 2.2], 5);
+        expect(traceOut.marker.size).toEqual([0.1, 0.2]);
+        expect(traceOut.marker.color).toEqual([10, 4]);
+        expect(traceOut.marker.opacity).toEqual([0.6, 'boo']);
+        expect(traceOut.marker.line.color).toEqual(['the end', 3.3]);
+        expect(traceOut.marker.line.width).toEqual([4, 1]);
+    });
+
+    it('handles all funcs except sum for date data', function() {
+        // weird cases handled in another test
+        Plotly.newPlot(gd, [{
+            x: ['2001-01-01', '', '2001-01-03', '2001-01-05', '2001-01-07'],
+            y: ['1995-01-15', '2005-03-15', '1990-12-23', '2001-01-01', 'not a date'],
+            text: ['2001-01-01 12:34', '2001-01-01 12:35', '2001-01-01 12:36', '2001-01-01 12:37', ''],
+            hovertext: ['a', '2001-01-02', '2001-01-03', '2001-01-04', '2001-01-05'],
+            customdata: ['2001-01', 'b', '2001-03', '2001-04', '2001-05'],
+            transforms: [{
+                type: 'aggregate',
+                // groups can be any type, but until we implement binning they
+                // will always compare as strings = so 1 === '1' === 1.0 !== '1.0'
+                groups: [1, 2, '1', 1.0, 1],
+                aggregations: [
```

Comment:
This would be a case where `Lib.keyedContainer` would be usable, though since we probably don't modify it internally, it probably serves no purpose. But for the workspace it could be possible to import it from plotly lib. It basically just helps you use this arrangement as if it's a direct key-value store.